File: app/schemas.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

MonitorType = Literal["http", "tcp", "dns"]
MonitorSeverity = Literal["critical", "high", "medium", "low"]
IncidentState = Literal["open", "acknowledged", "resolved"]
IncidentEventType = Literal["opened", "acknowledged", "resolved", "note_added"]
UserRole = Literal["user", "programmer", "admin"]


class MonitorCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    type: MonitorType
    service: str = Field(min_length=1, max_length=120)
    environment: str = Field(min_length=1, max_length=80)
    owner: str = Field(min_length=1, max_length=120)
    severity: MonitorSeverity = "medium"
    runbook_url: str | None = Field(default=None, max_length=500)
    target: str = Field(min_length=1, max_length=500)
    interval_seconds: int = Field(default=60, ge=5, le=3600)
    timeout_seconds: int = Field(default=5, ge=1, le=60)
    incident_threshold: int = Field(default=3, ge=1, le=20)
    retries: int = Field(default=0, ge=0, le=10)
    http_keyword: str | None = Field(default=None, max_length=500)
    enabled: bool = True


class MonitorUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    type: MonitorType | None = None
    service: str | None = Field(default=None, min_length=1, max_length=120)
    environment: str | None = Field(default=None, min_length=1, max_length=80)
    owner: str | None = Field(default=None, min_length=1, max_length=120)
    severity: MonitorSeverity | None = None
    runbook_url: str | None = Field(default=None, max_length=500)
    target: str | None = Field(default=None, min_length=1, max_length=500)
    interval_seconds: int | None = Field(default=None, ge=5, le=3600)
    timeout_seconds: int | None = Field(default=None, ge=1, le=60)
    incident_threshold: int | None = Field(default=None, ge=1, le=20)
    retries: int | None = Field(default=None, ge=0, le=10)
    http_keyword: str | None = Field(default=None, max_length=500)
    enabled: bool | None = None


class MonitorOut(BaseModel):
    id: int
    name: str
    type: str
    service: str
    environment: str
    owner: str
    severity: MonitorSeverity
    runbook_url: str | None
    target: str
    interval_seconds: int
    timeout_seconds: int
    incident_threshold: int
    retries: int
    http_keyword: str | None
    enabled: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class IncidentEventOut(BaseModel):
    id: int
    incident_id: int
    event_type: IncidentEventType
    actor: str
    note: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class IncidentOut(BaseModel):
    id: int
    monitor_id: int
    monitor_name: str | None = None
    state: IncidentState
    opened_at: datetime
    resolved_at: datetime | None = None
    failure_count: int
    last_error: str | None = None
    service: str
    environment: str
    owner: str
    severity: MonitorSeverity
    runbook_url: str | None = None
    timeline: list[IncidentEventOut] | None = None

    model_config = {"from_attributes": True}


class IncidentNoteCreate(BaseModel):
    note: str = Field(min_length=1, max_length=4000)


class CheckRunOut(BaseModel):
    id: int
    monitor_id: int
    monitor_name: str | None = None
    started_at: datetime
    duration_ms: int
    attempts: int
    success: bool
    status_code: int | None = None
    error: str | None = None

    model_config = {"from_attributes": True}


class WindowOut(BaseModel):
    minutes: int
    start: str
    end: str


class LatencyOut(BaseModel):
    p50: int | None = None
    p95: int | None = None
    min: int | None = None
    max: int | None = None


class MonitorBriefOut(BaseModel):
    id: int
    name: str
    type: str
    service: str
    environment: str
    owner: str
    severity: MonitorSeverity
    runbook_url: str | None = None
    target: str
    enabled: bool
    interval_seconds: int
    timeout_seconds: int


class IncidentBriefOut(BaseModel):
    id: int
    state: IncidentState
    opened_at: str
    failure_count: int
    last_error: str | None = None


class MaintenanceActiveOut(BaseModel):
    active: bool
    ends_at: datetime | None = None
    reason: str | None = None


class MonitorStatusOut(BaseModel):
    monitor: MonitorBriefOut
    status: Literal["up", "down", "unknown", "maintenance"]
    uptime_pct: float | None = None
    latency_ms: LatencyOut
    maintenance: MaintenanceActiveOut
    last_run: CheckRunOut | None = None
    open_incident: IncidentBriefOut | None = None


class OverviewOut(BaseModel):
    window: WindowOut
    monitors: list[MonitorStatusOut]


class StatusOut(BaseModel):
    generated_at: str
    overall: Literal["up", "degraded", "down"]
    window: WindowOut
    monitors: list[MonitorStatusOut]
    open_incidents: list[IncidentOut]


class MaintenanceCreate(BaseModel):
    monitor_id: int | None = None  # null means global
    starts_at: datetime
    ends_at: datetime
    reason: str | None = None


class MaintenanceOut(BaseModel):
    id: int
    monitor_id: int | None = None
    starts_at: datetime
    ends_at: datetime
    reason: str | None = None
    created_at: datetime
    model_config = {"from_attributes": True}


class AuditEventOut(BaseModel):
    id: int
    created_at: datetime
    actor: str
    action: str
    resource_type: str
    resource_id: int
    summary_json: dict[str, Any]

    model_config = {"from_attributes": True}


class AuthLoginIn(BaseModel):
    email: str = Field(min_length=3, max_length=320)
    password: str = Field(min_length=8, max_length=200)


class UserCreate(BaseModel):
    email: str = Field(min_length=3, max_length=320)
    display_name: str = Field(min_length=1, max_length=120)
    password: str = Field(min_length=8, max_length=200)
    role: UserRole = "user"
    is_active: bool = True


class UserUpdate(BaseModel):
    email: str | None = Field(default=None, min_length=3, max_length=320)
    display_name: str | None = Field(default=None, min_length=1, max_length=120)
    password: str | None = Field(default=None, min_length=8, max_length=200)
    role: UserRole | None = None
    is_active: bool | None = None


class UserOut(BaseModel):
    id: int
    email: str
    display_name: str
    role: UserRole
    is_active: bool
    created_at: datetime
    updated_at: datetime
    last_login_at: datetime | None = None

    model_config = {"from_attributes": True}


class AuthMeOut(BaseModel):
    id: int | None = None
    email: str
    display_name: str
    role: UserRole
    is_active: bool
    auth_method: Literal["session", "api_key"]
    last_login_at: datetime | None = None
    session_expires_at: datetime | None = None

File: frontend/src/lib/schemas.ts
import { z } from "zod";

export const userRoleSchema = z.enum(["user", "programmer", "admin"]);

export const incidentEventSchema = z.object({
  id: z.number(),
  incident_id: z.number(),
  event_type: z.enum(["opened", "acknowledged", "resolved", "note_added"]),
  actor: z.string(),
  note: z.string().nullable(),
  created_at: z.string(),
});

export const checkRunSchema = z.object({
  id: z.number(),
  monitor_id: z.number(),
  monitor_name: z.string().nullable().optional(),
  started_at: z.string(),
  duration_ms: z.number(),
  attempts: z.number(),
  success: z.boolean(),
  status_code: z.number().nullable(),
  error: z.string().nullable(),
});

const latencySchema = z.object({
  p50: z.number().nullable(),
  p95: z.number().nullable(),
  min: z.number().nullable(),
  max: z.number().nullable(),
});

const monitorBriefSchema = z.object({
  id: z.number(),
  name: z.string(),
  type: z.string(),
  service: z.string(),
  environment: z.string(),
  owner: z.string(),
  severity: z.enum(["critical", "high", "medium", "low"]),
  runbook_url: z.string().nullable(),
  target: z.string(),
  enabled: z.boolean(),
  interval_seconds: z.number(),
  timeout_seconds: z.number(),
});

const maintenanceSchema = z.object({
  active: z.boolean(),
  ends_at: z.string().nullable(),
  reason: z.string().nullable(),
});

const incidentBriefSchema = z.object({
  id: z.number(),
  state: z.enum(["open", "acknowledged", "resolved"]),
  opened_at: z.string(),
  failure_count: z.number(),
  last_error: z.string().nullable(),
});

export const monitorStatusSchema = z.object({
  monitor: monitorBriefSchema,
  status: z.enum(["up", "down", "unknown", "maintenance"]),
  uptime_pct: z.number().nullable(),
  latency_ms: latencySchema,
  maintenance: maintenanceSchema,
  last_run: checkRunSchema.nullable(),
  open_incident: incidentBriefSchema.nullable(),
});

export const incidentSchema = z.object({
  id: z.number(),
  monitor_id: z.number(),
  monitor_name: z.string().nullable().optional(),
  state: z.enum(["open", "acknowledged", "resolved"]),
  opened_at: z.string(),
  resolved_at: z.string().nullable(),
  failure_count: z.number(),
  last_error: z.string().nullable(),
  service: z.string(),
  environment: z.string(),
  owner: z.string(),
  severity: z.enum(["critical", "high", "medium", "low"]),
  runbook_url: z.string().nullable(),
  timeline: z.array(incidentEventSchema).nullable().optional(),
});

const windowSchema = z.object({
  minutes: z.number(),
  start: z.string(),
  end: z.string(),
});

export const statusSchema = z.object({
  generated_at: z.string(),
  overall: z.enum(["up", "degraded", "down"]),
  window: windowSchema,
  monitors: z.array(monitorStatusSchema),
  open_incidents: z.array(incidentSchema),
});

export const summarySchema = z.object({
  monitors: z.object({
    total: z.number(),
    enabled: z.number(),
  }),
  incidents: z.object({
    open_total: z.number(),
    open_actionable: z.number(),
    latest_total_open: z.array(
      z.object({
        id: z.number(),
        monitor_id: z.number(),
        monitor_name: z.string().nullable().optional(),
        state: z.enum(["open", "acknowledged", "resolved"]),
        opened_at: z.string(),
        failure_count: z.number(),
        last_error: z.string().nullable(),
        service: z.string(),
        environment: z.string(),
        owner: z.string(),
        severity: z.enum(["critical", "high", "medium", "low"]),
        runbook_url: z.string().nullable(),
      }),
    ),
    latest_actionable_open: z.array(
      z.object({
        id: z.number(),
        monitor_id: z.number(),
        monitor_name: z.string().nullable().optional(),
        state: z.enum(["open", "acknowledged", "resolved"]),
        opened_at: z.string(),
        failure_count: z.number(),
        last_error: z.string().nullable(),
        service: z.string(),
        environment: z.string(),
        owner: z.string(),
        severity: z.enum(["critical", "high", "medium", "low"]),
        runbook_url: z.string().nullable(),
      }),
    ),
  }),
});

export const overviewSchema = z.object({
  window: windowSchema,
  monitors: z.array(monitorStatusSchema),
});

export const monitorStatsSchema = z.object({
  monitor: monitorBriefSchema,
  window: windowSchema,
  runs: z.object({
    total: z.number(),
    success: z.number(),
    failure: z.number(),
    uptime_pct: z.number().nullable(),
  }),
  latency_ms: latencySchema,
});

export const maintenanceSchemaOut = z.object({
  id: z.number(),
  monitor_id: z.number().nullable(),
  starts_at: z.string(),
  ends_at: z.string(),
  reason: z.string().nullable(),
  created_at: z.string(),
});

export const monitorSchema = z.object({
  id: z.number(),
  name: z.string(),
  type: z.string(),
  service: z.string(),
  environment: z.string(),
  owner: z.string(),
  severity: z.enum(["critical", "high", "medium", "low"]),
  runbook_url: z.string().nullable(),
  target: z.string(),
  interval_seconds: z.number(),
  timeout_seconds: z.number(),
  incident_threshold: z.number(),
  retries: z.number(),
  http_keyword: z.string().nullable(),
  enabled: z.boolean(),
  created_at: z.string(),
});

export const versionSchema = z.object({
  version: z.string(),
  commit: z.string(),
  built_at: z.string(),
});

export const authMeSchema = z.object({
  id: z.number().nullable(),
  email: z.string(),
  display_name: z.string(),
  role: userRoleSchema,
  is_active: z.boolean(),
  auth_method: z.enum(["session", "api_key"]),
  last_login_at: z.string().nullable(),
  session_expires_at: z.string().nullable(),
});

export const userSchema = z.object({
  id: z.number(),
  email: z.string(),
  display_name: z.string(),
  role: userRoleSchema,
  is_active: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
  last_login_at: z.string().nullable(),
});

export const auditEventSchema = z.object({
  id: z.number(),
  created_at: z.string(),
  actor: z.string(),
  action: z.string(),
  resource_type: z.string(),
  resource_id: z.number(),
  summary_json: z.record(z.string(), z.unknown()),
});

export type CheckRun = z.infer<typeof checkRunSchema>;
export type Incident = z.infer<typeof incidentSchema>;
export type Monitor = z.infer<typeof monitorSchema>;
export type MonitorStats = z.infer<typeof monitorStatsSchema>;
export type MonitorStatus = z.infer<typeof monitorStatusSchema>;
export type OverviewResponse = z.infer<typeof overviewSchema>;
export type StatusResponse = z.infer<typeof statusSchema>;
export type SummaryResponse = z.infer<typeof summarySchema>;
export type VersionResponse = z.infer<typeof versionSchema>;
export type AuthMe = z.infer<typeof authMeSchema>;
export type User = z.infer<typeof userSchema>;
export type AuditEvent = z.infer<typeof auditEventSchema>;
